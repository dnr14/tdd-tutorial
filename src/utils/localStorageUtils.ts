export const _TODO_KEY = 'todo_list';
