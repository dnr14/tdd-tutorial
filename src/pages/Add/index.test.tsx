import { screen, render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { Router, useLocation } from 'react-router-dom';
import { Add } from '.';
import { createMemoryHistory } from 'history';
import { _TODO_KEY } from 'utils/localStorageUtils';

const PLACE_HOLDER = '할 일을 입력해주세요.';

beforeEach(() => {
  localStorage.clear();
});

const TestCompo = () => {
  const { pathname } = useLocation();
  return <div>{pathname}</div>;
};

const HOC = ({ path = '/' }: { path?: string }) => {
  const history = createMemoryHistory();
  history.push(path);
  return (
    <Router history={history}>
      <Add />
      <TestCompo />
    </Router>
  );
};

describe('<Add>', () => {
  it('Add 컴포넌트 랜더링', () => {
    const { container } = render(<HOC />);
    const placeholder = screen.getByPlaceholderText(PLACE_HOLDER);
    expect(placeholder).toBeInTheDocument();

    const addButton = screen.getByText('추가');
    expect(addButton).toBeInTheDocument();

    fireEvent.change(placeholder, { target: { value: 'todo1' } });
    fireEvent.click(addButton);
    expect(localStorage.getItem(_TODO_KEY)).toBe('["todo1"]');

    expect(container).toMatchSnapshot();
  });

  it('Add 페이지 이동 후 할 일 추가 후 홈으로 돌아간다.', () => {
    render(<HOC path="/add" />);

    const url = screen.getByText('/add');
    expect(url).toBeInTheDocument();

    const placeholder = screen.getByPlaceholderText(PLACE_HOLDER);
    expect(placeholder).toBeInTheDocument();
    fireEvent.change(placeholder, { target: { value: 'todo1' } });
    fireEvent.click(screen.getByText('추가'));
    expect(url.textContent).toBe('/');
  });

  it('인풋태그에 값이 없으면 아이템이 추가되면 안된다.', () => {
    render(<HOC path="/add" />);

    const url = screen.getByText('/add');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('추가');
    fireEvent.click(addButton);

    expect(localStorage.getItem(_TODO_KEY)).toBe(null);
    expect(url.textContent).toBe('/add');
  });
});
