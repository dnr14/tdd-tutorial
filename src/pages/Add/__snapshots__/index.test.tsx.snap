// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Add> Add 컴포넌트 랜더링 1`] = `
.c2 {
  text-align: center;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  background-color: #304ffe;
}

.c2:hover {
  background-color: #1e40ff;
}

.c2:active {
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
}

.c3 {
  color: #fff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  padding: 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="할 일을 입력해주세요."
      type="text"
    />
    <div
      class="c2"
      data-testid="button-container"
    >
      <div
        class="c3"
      >
        추가
      </div>
    </div>
  </div>
  <div>
    /
  </div>
</div>
`;
